oxidauth-usecases/src/auth/authenticate_or_register.rs

```diff
@@ -0,0 +1,227 @@
+    authorities: T,
+}
+
+impl<A, M, P, R, S, T, UI, U, UU> AuthenticateOrRegisterUseCase<A, M, P, R, S, T, UI, U, UU>
+where
+    A: InsertUserAuthorityQuery,
+    M: SelectMostRecentPrivateKeyQuery,
+    P: PermissionTreeQuery,
+    R: InsertRefreshTokenQuery,
+    S: SelectTOTPSecrețByUserIdQuery,
+    T: SelectAuthorityByClientKeyQuery,
+    UI: InsertUserQuery,
+    U: SelectUserAuthoritiesByAuthorityIdAndUserIdentifierQuery,
+    UU: SelectUserByIdQuery,
+{
+    pub fn new(
+        authenticate: Arc<AuthenticateUseCase<T, U, P, M, R, S, UU>>,
+        register: Arc<RegisterUseCase<T, UI, A, P, M, R>>,
+        authorities: T,
+    ) -> Self {
+        Self {
+            authenticate,
+            register,
+            authorities,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a, A, M, P, R, S, T, UI, U, UU> Service<&'a AuthenticateOrRegisterParams>
+    for AuthenticateOrRegisterUseCase<A, M, P, R, S, T, UI, U, UU>
+where
+    A: InsertUserAuthorityQuery,
+    M: SelectMostRecentPrivateKeyQuery,
+    P: PermissionTreeQuery,
+    R: InsertRefreshTokenQuery,
+    S: SelectTOTPSecrețByUserIdQuery,
+    T: SelectAuthorityByClientKeyQuery,
+    UI: InsertUserQuery,
+    U: SelectUserAuthoritiesByAuthorityIdAndUserIdentifierQuery,
+    UU: SelectUserByIdQuery,
+{
+    type Response = AuthenticateOrRegisterResponse;
+    type Error = BoxedError;
+
+    #[tracing::instrument(name = "authenticate_or_register_usecase", skip(self))]
+    async fn call(
+        &self,
+        params: &'a AuthenticateOrRegisterParams,
+    ) -> Result<Self::Response, Self::Error> {
+        // GET AUTHORITY
+        let Some(authority) = self
+            .authorities
+            .call(&FindAuthorityByClientKey {
+                client_key: params.client_key,
+            })
+            .await?
+        else {
+            return Err(format!("could not find authority by client key").into());
+        };
+
+        // GET PARAMS
+        let authority_params: AuthorityParams = authority.params.try_into()?;
+        let authenticate_params: OAuth2AuthenticateParams = params
+            .params
+            .clone()
+            .try_into()?;
+
+        let json = GoogleExchangeTokenReq {
+            code: authenticate_params
+                .code
+                .to_owned(),
+            client_id: authority_params.oauth2_id,
+            client_secret: authority_params.oauth2_secret,
+            redirect_uri: authority_params
+                .redirect_uri
+                .to_string(),
+            grant_type: String::from("authorization_code"),
+        };
+
+        let exchange: GoogleExchangeTokenRes = reqwest::Client::new()
+            .post("https://oauth2.googleapis.com/token")
+            .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
+            .form(&json)
+            .send()
+            .await
+            .map_err(|err| err.to_string())?
+            .json()
+            .await
+            .map_err(|err| err.to_string())?;
```

Comment:
```suggestion
            .map_err(ToString::to_string)?
            .json()
            .await
            .map_err(ToString::to_string)?;
```

Reply:
A closure == function pointer. Just a bit cleaner/more semantic.

Reply:
This causes errors, we should review